sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -490,7 +490,12 @@ private class LiveExecutionData(val executionId: Long) extends LiveEntity {
   var details: String = null
   var physicalPlanDescription: String = null
   var modifiedConfigs: Map[String, String] = _
-  var metrics = collection.Seq[SQLPlanMetric]()
+  private var _metrics = collection.Seq[SQLPlanMetric]()
+  def metrics: collection.Seq[SQLPlanMetric] = _metrics
+  // This mapping is shared across all LiveStageMetrics instances associated with
+  // this LiveExecutionData, helping to reduce memory overhead by avoiding waste
+  // from separate immutable maps with largely overlapping sets of entries.
+  val metricAccumulatorIdToMetricType = new ConcurrentHashMap[Long, String]()
```

Comment:
AFAIK, SQLAppStatusListener only works in single thread. Should we only use Map here? please tell me detail if I thought missing something.

Reply:
Our Spark fork contains a codepath that performs multi-threaded access to this data structure, so that's why I used a thread-safe data structure here.

If you would prefer that I avoid leaking those proprietary thread-safety concerns into OSS Apache Spark then I would be glad to refactor this to use a `mutable.Map` instead and keep the thread-safe version in our fork.

Reply:
Curious why this is diverging ? How is it used ? Is it something that might be relevant to Apache Spark as a functional enhancement ?

Reply:
The codepath in question isn't of general relevance and, as it turns out, might be deprecated soon. Given this, I have updated this PR to use a `mutable.HashMap` (which simplifies the rest of the code and addresses your other review comment).